starlette/types.py

```diff
@@ -1,9 +1,17 @@
+import sys
 import typing
 
-Scope = typing.MutableMapping[str, typing.Any]
-Message = typing.MutableMapping[str, typing.Any]
+if sys.version_info >= (3, 10):  # pragma: no cover
+    from typing import TypeAlias
+else:  # pragma: no cover
+    from typing_extensions import TypeAlias
 
-Receive = typing.Callable[[], typing.Awaitable[Message]]
-Send = typing.Callable[[Message], typing.Awaitable[None]]
+Scope: TypeAlias = typing.MutableMapping[str, typing.Any]
+Message: TypeAlias = typing.MutableMapping[str, typing.Any]
 
-ASGIApp = typing.Callable[[Scope, Receive, Send], typing.Awaitable[None]]
+Receive: TypeAlias = typing.Callable[[], typing.Awaitable[Message]]
+Send: TypeAlias = typing.Callable[[Message], typing.Awaitable[None]]
+
+ASGIApp: TypeAlias = typing.Callable[[Scope, Receive, Send], typing.Awaitable[None]]
+
+MiddewareType: TypeAlias = ASGIApp
```

Comment:
https://github.com/encode/starlette/blob/337ae243b10d3c20db53b77bc3c7718bb4f1a164/starlette/applications.py#L103-L104

Considering this code, should we use
```python
class MiddlewareType(Protocol):
    def __call__(
        self,
        app: ASGIApp,
        **options: Any
    ) -> ASGIApp:
        ...
```

Reply:
If you look at the [`__iter__`](https://github.com/encode/starlette/blob/337ae243b10d3c20db53b77bc3c7718bb4f1a164/starlette/middleware/__init__.py#L9-L11) method of the `Middleware` class, you will see that it returns a tuple whose first element is the middleware class

Reply:
`cls` is `MiddlewareType`.

Reply:
Yes indeed sorry, there's some confusion here between the `__init__` and the `__call__` method. When doing `app = cls(app=app, **options)`, we are actually instantiating the middleware, not using the `__call__` method.

Defining the `MiddlewareType` as a `Callable` or as a `Protocol` implementing `__call__` is equivalent, as `Callable` is an ABC implementing `__call__`

Reply:
This is a ASGI middleware that can be used in starlette.

```python
class ASGIMiddleware:
    def __init__(app: ASGIApp, p0: str, p1: int) -> None:
        ...
    async def __call__(self, scope, receive, send) -> None:
        ...
```

This also can be used in starlette.

```python
def asgi_middleware(app, p0: str, p1: int) -> ASGIApp:
    def _asgi(scope, receive, send) -> None:
        ...
    return _asgi
```

But this can't be used in starlette. Even though `P` passes the type check you're giving now.

```python
class P:
    def __init__(self) -> None: ...
    def __call__(self, scope, receive, send) -> None: ...
```

Reply:
If we are going to write Type Annotation, then the parameter type needs to be as precise as possible, otherwise it will not bring any benefit.

Reply:
> https://github.com/encode/starlette/blob/337ae243b10d3c20db53b77bc3c7718bb4f1a164/starlette/applications.py#L103-L104
> 
> Considering this code, should we use
> 
> ```python
> class MiddlewareType(Protocol):
>     def __call__(
>         self,
>         app: ASGIApp,
>         **options: Any
>     ) -> ASGIApp:
>         ...
> ```

In other places where `typing.Type[MiddlewareType]` is now used, just change to `MiddlewareType`.

Reply:
> If we are going to write Type Annotation, then the parameter type needs to be as precise as possible, otherwise it will not bring any benefit.

**Without** having read any of the other discussion here, I would tend to agree, incorrect type annotations are worse than none.

Reply:
Too bad protocols aren't supporting the `__init__` method (for reference: https://bugs.python.org/issue44807, where a use case similar to this one is discussed). If I understand correctly, by using your suggestion the `__call__` method is actually used as the `__init__` method? So we get the `__init__` method correctly typed but we loose information on the real async `__call__` :/

Reply:
> Too bad protocols aren't supporting the `__init__` method (for reference: https://bugs.python.org/issue44807, where a use case similar to this one is discussed). If I understand correctly, by using your suggestion the `__call__` method is actually used as the `__init__` method? So we get the `__init__` method correctly typed but we loose information on the real async `__call__` :/

No, this is just a way of specifying the name of the `Callable` parameter. In fact it is not limited to classes.

Reply:
I'm not sure I'm following you on this one. What I understand is that middlewares should have the following implementation:

```python
class MyMiddleware:
    def __init__(self, app: ASGIApp, **options: Any) -> None:
        ...

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        ...
```

So considering the following code: https://github.com/encode/starlette/blob/337ae243b10d3c20db53b77bc3c7718bb4f1a164/starlette/applications.py#L103-L104

`cls` is actually the class of the middleware, not an instance. Hence calling `cls(app=app, **options)` is actually calling the `__init__` method of the middleware and instantiating it. The problem I see we are facing is that there's no way to specify the `__init__` method signature in a `Protocol` definition

Reply:
```python
class MyMiddleware:
    def __init__(self, app: ASGIApp, **options: Any) -> None:
        ...

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        ...
```

This class passes the following type checks.

```python
class MiddlewareType(Protocol):
    def __call__(
        self,
        app: ASGIApp,
        **options: Any
    ) -> ASGIApp:
        ...
```

Reply:
It might, but imo this is not the way to go:
- You are using the `__call__` method as a way to define the signature of the `__init__` method, which is confusing
- the `MiddlewareType` you created does not have any signature for the real async `__call__` method that should be defined on a middleware class

Reply:
If you don't know how to use typing.Protocol, please consider using mypy to give it a try. 😊

Reply:
I know how Protocols work, but you haven't explained why you are using these parameters for the `__call__` method (see my answers above). As I understand currently, you are using the `__call__` method of your protocol as a way to define the `__init__` method of a middleware, right?

Reply:
Will continue discussion in main thread for clarity.